aten/src/ATen/core/jit_type.h

```diff
@@ -324,25 +324,119 @@ inline c10::optional<T> merge_primitive(
+// However, if we knew the WxH dimensions were contiguous, 
+// we can pretend like we only have a single dimension, simplifying the indexing logic. 
+// This can be performed even if the dimensions are transposed, 
+// as long as a, b, and c are transposed in the same way. 
+// We’d like to have the compiler be able to do this dimensionality reduction, 
+// but simply knowing sizes is not enough. 
+// We can extend profiling to also record stride information. 
+// Rather than recording specific strides, 
+// we can simply order the strides from smallest to largest with `stride_indices`
+// A contiguity marker on the smallest stride (c0) indicates the stride is precisely 1, 
+// otherwise a contiguity marker means that $stride_n = size_{n-1}*stride_{n-1}$
+struct CAFFE2_API Stride {
+  Stride() {}
+  Stride(const c10::optional<size_t>& stride_index, const c10::optional<bool>& contiguous, const c10::optional<size_t>& stride)
+  : stride_index_(stride_index)
+  , contiguous_(contiguous)
+  , stride_(stride) {}
+
+
+  bool operator ==(const Stride &b) const { 
+    return stride_index_ == b.stride_index_ && 
+      contiguous_ == b.contiguous_ && 
+      stride_ == b.stride_; 
+  }
+
+  c10::optional<size_t> stride_index_;
+  c10::optional<bool> contiguous_;
+  c10::optional<size_t> stride_;
+};
+
+template <>
+inline c10::optional<Stride> merge_primitive(
+    const c10::optional<Stride>& a,
+    const c10::optional<Stride>& b) {
+
+  c10::optional<Stride> left = a;
+  c10::optional<Stride> right = b;
+  if (!left.has_value()) {
+    left = {Stride()};
+  }
+  if (!right.has_value()) {
+    right = {Stride()};
+  }
+
+  auto merged_index = merge_primitive(left->stride_index_, right->stride_index_);
+  auto merged_cont = merge_primitive(left->contiguous_, right->contiguous_);
+  auto merged_stride = merge_primitive(left->stride_, right->stride_);
+  auto r = Stride (merged_index, merged_cont, merged_stride);
+  // normalize
+  if (!r.stride_index_.has_value() && !r.contiguous_.has_value() && !r.stride_.has_value())
+  {
+    return c10::optional<Stride>{};
+  }
+
+  return r;  
+}
+
+struct CAFFE2_API ShapeSymbol {
+    
+    // needed for use in `std::map`
+    ShapeSymbol(): value_(-1) {}
+    // is this symbol a fixed/static dimension
+    bool is_static() const { return value_ >= 0; };  
+    bool operator ==(const ShapeSymbol &b) const { return value_ == b.value_; }
+    bool operator <(const ShapeSymbol &b) const { return value_ < b.value_; }
+
+    static ShapeSymbol fromStaticSize(int64_t val) {return ShapeSymbol(val); }
+    int64_t static_size() const {
+      TORCH_CHECK(is_static());
+      return value_;
+    };
+
+    static ShapeSymbol newSymbol () {return -(++num_symbols); };
+    friend std::ostream& operator<<(std::ostream& os, const ShapeSymbol& s);
+    private:
+      ShapeSymbol(int64_t val): value_(val) {}
+      int64_t value_;
+      static std::atomic<size_t> num_symbols;
+ 
+};
+
+
+template <typename T>
```

Comment:
Both Stride, and ShapeSymbol now have their own internal ways of representing not present values. So VaryingShape is going to get confusing because of the possibilities involved. It would be good to try to simply the API surface here. For instance does `sizes()` need to still return a list of optional ints? Or do we just need concrete sizes, and sizes_properaties() (i.e. the symbolic shape-returning version).

Reply:
I initially liked the consistency of having `c10::optional` to denote the `unknowness` but it's getting a bit clunky and awkward especially with `Stride`s. 
We could probably switch to the following API:
```cpp
c10::optional<std::vector<ShapeSymbol> symbolic_sizes();